examples/miral-custom-decorations/miral_custom_decorations.cpp

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#include "mir/server.h"
+#include "mir/main_loop.h" // Needed for the conversion of `the_main_loop` to an executor
+
+#include "floating_window_manager.h"
+#include "miral/decoration/decoration_manager_adapter.h"
+#include "miral/runner.h"
+#include "miral/set_window_management_policy.h"
+#include "user_decoration_manager_example.h"
+#include <memory>
+#include <miral/display_configuration_option.h>
+#include <miral/external_client.h>
+#include <miral/window_management_options.h>
+#include <miral/append_event_filter.h>
+#include <miral/internal_client.h>
+#include <miral/command_line_option.h>
+#include <miral/cursor_theme.h>
+#include <miral/decorations.h>
+#include <miral/keymap.h>
+#include <miral/toolkit_event.h>
+#include <miral/x11_support.h>
+#include <miral/wayland_extensions.h>
+
+#include <xkbcommon/xkbcommon-keysyms.h>
+
+#include <cstring>
+
+namespace
+{
+struct ConfigureDecorations
+{
+    miral::Decorations const decorations{[]
+        {
+            if (auto const strategy = getenv("MIRAL_SHELL_DECORATIONS"))
+            {
+                if (strcmp(strategy, "always-ssd") == 0) return miral::Decorations::always_ssd();
+                if (strcmp(strategy, "prefer-ssd") == 0) return miral::Decorations::prefer_ssd();
+                if (strcmp(strategy, "always-csd") == 0) return miral::Decorations::always_csd();
+            }
+            return miral::Decorations::prefer_csd();
+        }()};
+
+    void operator()(mir::Server& s) const
+    {
+        decorations(s);
+    }
+};
+
+struct CustomDecorations
+{
+    CustomDecorations() = default;
+
+    void operator()(mir::Server& s) const
+    {
+        s.add_pre_init_callback(
```

Comment:
User code should not be accessing the Server in this way (it isn't a public header)